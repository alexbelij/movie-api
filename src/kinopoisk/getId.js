/* @flow */

import cheerio from 'cheerio';
import R from 'ramda';
import similarity from 'similarity';

import connector from './connector';

type Query = {
  title: string,
  year?: number,
  countries?: Array<string>,
  isTvShow?: boolean,
};
type SearchResult = {
  id: number,
  title: string,
  countries: Array<string>,
  year: number,
};

const scrapeResults = (
  query: Query,
  html: string,
): Array<SearchResult> => {
  const $ = cheerio.load(html);

  return $(`.film-snippet_type_${query.isTvShow ? 'show' : 'movie'}`)
    .get()
    .map((snippet: any) => {
      const titleNode = $(snippet).find('.film-snippet__title-link');
      const id = parseInt(
        R.nth(1, titleNode.attr('href').match(/\/film\/(.+)\//)),
        10,
      );

      if (isNaN(id)) return null;

      const infoNode = $(snippet).find('.film-snippet__info');
      const title = titleNode.text().replace(/\s\s+/g, ' ');
      const countries = R.pipe(
        R.split(','),
        R.map(R.trim),
        R.filter(
          (country: ?string) => !!country && isNaN(parseInt(country, 10)),
        ),
      )(infoNode.text());
      const year = parseInt(R.head(infoNode.text().match(/\d+/)), 10);

      return { id, title, countries, year };
    });
};

const filterResults = (
  query: Query,
  results: Array<?SearchResult>,
): Array<SearchResult> => R.filter(R.allPass([
  Boolean,
  ({ title }: SearchResult) => similarity(title, query.title) >= 0.8,
  (query.year
    ? ({ year }: SearchResult) => year === query.year
    : R.always(true)),
  (query.countries
    ? ({ countries }: SearchResult) => R.equals(
        R.intersection(countries, query.countries).length,
        // $FlowFixMe: countries is always available here
        query.countries.length,
      )
    : R.always(true)),
]))(results);

const getId = async (query: Query): Promise<?number> => {
  const html = await connector.htmlGet('search', { text: query.title });

  return R.pipe(
    R.curry(scrapeResults)(query),
    R.curry(filterResults)(query),
    R.sortBy(({ title }: SearchResult) => similarity(title, query.title)),
    R.head,
    R.prop('id'),
  )(html);
};

export {
  scrapeResults as __scrapeResults,
  filterResults as __filterResults,
};
export default getId;
