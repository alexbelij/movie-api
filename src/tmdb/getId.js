/* @flow */

import R from 'ramda';

import connector from './connector';

type Query = {
  title?: string,
  year?: number,
  imdbId?: string,
  isTvShow?: boolean,
};

const findByImdbId = async (query: Query): Promise<?number> => {
  if (!query.imdbId) return null;

  const res = await connector.apiGet(
    `find/${query.imdbId}`,
    { external_source: 'imdb_id' },
  );

  const results = query.isTvShow
    ? R.propOr([], 'tv_results', res)
    : R.propOr([], 'movie_results', res);

  return R.pipe(R.head, R.prop('id'))(results);
};
const searchByTitle = async (query: Query): Promise<?number> => {
  if (!query.title) return null;

  const res = await connector.apiGet(
    `search/${query.isTvShow ? 'tv' : 'movie'}`,
    { query: query.title, year: query.year },
  );

  return R.pipe(R.prop('results'), R.head, R.prop('id'))(res);
};

const getId = async (query: Query): Promise<?number> => {
  if (query.imdbId) return findByImdbId(query);
  if (query.title) return searchByTitle(query);

  return null;
};

export {
  findByImdbId as __findByImdbId,
  searchByTitle as __searchByTitle,
};
export default getId;
