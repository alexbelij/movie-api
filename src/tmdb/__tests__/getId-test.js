/* @flow */

import getId from '../getId';

describe('tmdb/getId', () => {
  it('fetches movie id by imdb id', async () => {
    expect(await getId({ imdbId: 'tt2488496' })).toMatchSnapshot();
  });

  it('fetches movie id by title', async () => {
    expect(await getId({
      title: 'Star Wars: The Force Awakens',
    })).toMatchSnapshot();
  });

  it('fetches tv show id by imdb id', async () => {
    expect(await getId({
      imdbId: 'tt0944947',
      isTvShow: true,
    })).toMatchSnapshot();
  });

  it('fetches tv show id by title', async () => {
    expect(await getId({
      title: 'Game of Thrones',
      isTvShow: true,
    })).toMatchSnapshot();
  });
});
