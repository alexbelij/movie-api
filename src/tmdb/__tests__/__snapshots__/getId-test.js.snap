exports[`tmdb/getId fetches movie id by imdb id 1`] = `140607`;

exports[`tmdb/getId fetches movie id by title 1`] = `140607`;

exports[`tmdb/getId fetches tv show id by imdb id 1`] = `1399`;

exports[`tmdb/getId fetches tv show id by title 1`] = `1399`;
